assets/src/components/swingsView.tsx

```diff
@@ -0,0 +1,177 @@
+import { ByRouteId, Route, Swing } from "../schedule"
+import { selectVehicle, toggleSwingsView } from "../state"
+import {
+  formattedScheduledTime,
+  now,
+  serviceDaySeconds,
+} from "../util/dateTime"
+
+const SwingsView = (): ReactElement<HTMLElement> => {
+  const [, dispatch] = useContext(StateDispatchContext)
+  const [currentTime] = useState(now())
+  const swings = useSwings()
+
+  const activeSwings = swings
+    ? swings.filter((swing) => {
+        return swing.time > serviceDaySeconds(currentTime)
+      })
+    : []
+
+  const swingRunIds = uniq(
+    flatten(activeSwings.map((swing) => [swing.fromRunId, swing.toRunId]))
+  )
+
+  const { socket } = useContext(SocketContext)
+  const swingVehicles = useVehiclesForRunIds(socket, swingRunIds)
+  let swingVehiclesByRunId: ByRunId<VehicleOrGhost> = {}
+  if (swingVehicles) {
+    swingVehiclesByRunId = swingVehicles.reduce((map, vehicle) => {
+      if (vehicle.runId) {
+        return { ...map, [vehicle.runId]: vehicle }
+      } else {
+        return map
+      }
+    }, swingVehiclesByRunId)
+  }
+
+  const swingRouteIds = uniq(activeSwings.map((swing) => swing.fromRouteId))
+  const swingRoutes = useRoutes(swingRouteIds)
+  const swingRoutesById: ByRouteId<Route> = swingRoutes.reduce((map, route) => {
+    return { ...map, [route.id]: route }
+  }, {})
+
+  const hideMe = () => dispatch(toggleSwingsView())
+
+  return (
+    <div id="m-swings-view" className="m-swings-view">
+      <CloseButton onClick={hideMe} />
+      <div className="m-swings-view__header">Swings view</div>
+      <div className="m-swings-view__description">
+        Upcoming swings on your selected routes
+      </div>
+      {swings ? (
+        <SwingsTable
+          swings={activeSwings}
+          swingVehiclesByRunId={swingVehiclesByRunId}
+          swingRoutesById={swingRoutesById}
+        />
+      ) : (
+        <Loading />
+      )}
+    </div>
+  )
+}
+
+const SwingsTable = ({
+  swings,
+  swingVehiclesByRunId,
+  swingRoutesById,
+}: {
+  swings: Swing[]
+  swingVehiclesByRunId: ByRunId<VehicleOrGhost>
+  swingRoutesById: ByRouteId<Route>
+}): ReactElement<HTMLElement> => {
+  return (
+    <div className="m-swings-view__table-container">
+      <table className="m-swings-view__table">
+        <thead className="m-swings-view__table-header">
+          <tr>
+            <th className="m-swings-view__table-header-cell">Swing On Time</th>
+            <th className="m-swings-view__table-header-cell">
+              Swing On<div className="run-subheader">Run</div>
+            </th>
+            <th className="m-swings-view__table-header-cell swing-off">
+              Swing Off<div className="run-subheader">Run</div>
+              <div className="route-subheader">Route</div>
+            </th>
+            <th className="m-swings-view__table-header-cell">Vehicle No.</th>
+          </tr>
+        </thead>
+        <tbody>
+          {swings
+            .sort((swing1, swing2) => {
```

Comment:
let's move this `sort` out of the template and introduce a `sortedSwings` const, to avoid having logic in the JSX as far as possible